File: backend/app/schemas/fefo.py
"""
AVENZO Backend — FEFO (First-Expired, First-Out) Pydantic Schemas
"""

from uuid import UUID
from datetime import date
from typing import Optional, List
from pydantic import BaseModel, ConfigDict, Field


class FEFORankedBatchRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    batch_id: UUID
    batch_number: str
    product_id: UUID
    product_name: str
    sku: str
    has_expiry: bool
    manufacturing_date: Optional[date] = None
    expiry_date: Optional[date] = None
    days_to_expiry: Optional[int] = None
    expiry_status: str # SAFE, EXPIRING_SOON, CRITICAL, EXPIRED, N/A
    quantity_on_hand: int
    quantity_reserved: int
    quantity_available: int
    warehouse_id: UUID
    warehouse_name: str
    location_id: Optional[UUID] = None
    location_code: Optional[str] = None
    fefo_rank: int


class FEFOAllocationRequest(BaseModel):
    product_id: UUID
    requested_quantity: int = Field(..., gt=0, description="Quantity of units requested for picking")
    warehouse_id: Optional[UUID] = None


class FEFOBatchAllocationItem(BaseModel):
    batch_id: UUID
    batch_number: str
    manufacturing_date: Optional[date] = None
    expiry_date: Optional[date] = None
    days_to_expiry: Optional[int] = None
    expiry_status: str
    allocated_quantity: int
    quantity_available: int
    warehouse_id: UUID
    warehouse_name: str
    location_id: Optional[UUID] = None
    location_code: Optional[str] = None
    fefo_rank: int


class FEFOAllocationPlanResponse(BaseModel):
    product_id: UUID
    product_name: str
    sku: str
    requested_quantity: int
    allocated_total: int
    remaining_unallocated: int
    is_fully_allocated: bool
    allocations: List[FEFOBatchAllocationItem]


class FEFOVerificationRequest(BaseModel):
    product_id: UUID
    selected_batch_id: UUID
    requested_quantity: int = Field(..., gt=0)
    warehouse_id: Optional[UUID] = None
    override_reason: Optional[str] = Field(None, max_length=500)


class FEFOVerificationResponse(BaseModel):
    is_compliant: bool
    violation_detected: bool
    warning_message: Optional[str] = None
    selected_batch_id: UUID
    selected_expiry_date: Optional[date] = None
    earlier_available_batch_id: Optional[UUID] = None
    earlier_available_expiry_date: Optional[date] = None
    bypassed_earlier_quantity: int = 0
    audit_logged: bool = False

File: business-web/src/types/fefo.ts
/**
 * AVENZO Business Web — FEFO Intelligence Types
 */

export interface FEFORankedBatch {
  batch_id: string;
  batch_number: string;
  product_id: string;
  product_name: string;
  sku: string;
  has_expiry: boolean;
  manufacturing_date?: string;
  expiry_date?: string;
  days_to_expiry?: number;
  expiry_status: string; // SAFE, EXPIRING_SOON, CRITICAL, EXPIRED, N/A
  quantity_on_hand: number;
  quantity_reserved: number;
  quantity_available: number;
  warehouse_id: string;
  warehouse_name: string;
  location_id?: string;
  location_code?: string;
  fefo_rank: number;
}

export interface FEFOAllocationRequest {
  product_id: string;
  requested_quantity: number;
  warehouse_id?: string;
}

export interface FEFOBatchAllocationItem {
  batch_id: string;
  batch_number: string;
  manufacturing_date?: string;
  expiry_date?: string;
  days_to_expiry?: number;
  expiry_status: string;
  allocated_quantity: number;
  quantity_available: number;
  warehouse_id: string;
  warehouse_name: string;
  location_id?: string;
  location_code?: string;
  fefo_rank: number;
}

export interface FEFOAllocationPlan {
  product_id: string;
  product_name: string;
  sku: string;
  requested_quantity: number;
  allocated_total: number;
  remaining_unallocated: number;
  is_fully_allocated: boolean;
  allocations: FEFOBatchAllocationItem[];
}

export interface FEFOVerificationRequest {
  product_id: string;
  selected_batch_id: string;
  requested_quantity: number;
  warehouse_id?: string;
  override_reason?: string;
}

export interface FEFOVerificationResponse {
  is_compliant: boolean;
  violation_detected: boolean;
  warning_message?: string;
  selected_batch_id: string;
  selected_expiry_date?: string;
  earlier_available_batch_id?: string;
  earlier_available_expiry_date?: string;
  bypassed_earlier_quantity: number;
  audit_logged: boolean;
}
